### ISSUE TITLE
Question about performence on Egoschema benchmark

### ISSUE THREAD

**Issue by Xiaolong-RRL** [EvolvingLMMs-Lab/EgoLife#5]
Hi~

Thanks again for your great work!

I'd like to know what your approach to Egoschema benchmark is like. 

Does it set audio_path to None when loading_video?

I ran the following test code but it didn't work as well, specifically less than 40% accuracy on the Egoschema subset500 validation set:

```python
import argparse
import copy
import os
import re
import sys
import warnings
import json
from tqdm import tqdm
import numpy as np
import requests
import soundfile as sf
import torch
import torch.distributed as dist
import whisper
from decord import VideoReader, cpu
from egogpt.constants import (
    DEFAULT_IMAGE_TOKEN,
    DEFAULT_SPEECH_TOKEN,
    IGNORE_INDEX,
    IMAGE_TOKEN_INDEX,
    SPEECH_TOKEN_INDEX,
)
import uuid
from egogpt.conversation import SeparatorStyle, conv_templates
from egogpt.mm_utils import get_model_name_from_path, process_images
from egogpt.model.builder import load_pretrained_model
from PIL import Image
from scipy.signal import resample


def setup(rank, world_size):
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "12355"
    dist.init_process_group("gloo", rank=rank, world_size=world_size)


def load_video(video_path=None, audio_path=None, max_frames_num=16, fps=1):
    if audio_path is not None:
        speech, sample_rate = sf.read(audio_path)
        if sample_rate != 16000:
            target_length = int(len(speech) * 16000 / sample_rate)
            speech = resample(speech, target_length)
        if speech.ndim > 1:
            speech = np.mean(speech, axis=1)
        speech = whisper.pad_or_trim(speech.astype(np.float32))
        speech = whisper.log_mel_spectrogram(speech, n_mels=128).permute(1, 0)
        speech_lengths = torch.LongTensor([speech.shape[0]])
    else:
        speech = torch.zeros(3000, 128)
        speech_lengths = torch.LongTensor([3000])

    vr = VideoReader(video_path, ctx=cpu(0), num_threads=1)
    total_frame_num = len(vr)
    avg_fps = round(vr.get_avg_fps() / fps)
    frame_idx = [i for i in range(0, total_frame_num, avg_fps)]
    if max_frames_num > 0 and len(frame_idx) > max_frames_num:
        uniform_sampled_frames = np.linspace(
            0, total_frame_num - 1, max_frames_num, dtype=int
        )
        frame_idx = uniform_sampled_frames.tolist()
    video = vr.get_batch(frame_idx).asnumpy()
    return video, speech, speech_lengths


def split_text(text, keywords):
    pattern = "(" + "|".join(map(re.escape, keywords)) + ")"
    parts = re.split(pattern, text)
    parts = [part for part in parts if part]
    return parts


def main(args):
    warnings.filterwarnings("ignore")
    setup(0, 1)
    device_map = 'cuda:0' if torch.cuda.is_available() else 'cpu'
    model_name = args.model_path
    tokenizer, model, max_length = load_pretrained_model(
        args.model_path, device_map=device_map
    )
    model.eval()
    device = model.device

    outputs = []
    data_path=os.path.join(args.data_path, "subset_questions.json"  if args.eval_type == "subset" else "miniset_questions.json")
    shard_dataset = json.load(open(data_path, "r"))
    for line in tqdm(shard_dataset):
        video_name = line["google_drive_id"]
        idx = line["q_uid"]
        question = line["question"]
        a0 = line["option 0"]
        a1 = line["option 1"]
        a2 = line["option 2"]
        a3 = line["option 3"]
        a4 = line["option 4"]
        qs = f"Question: {question}\nOptions:\n(A) {a0}\n(B) {a1}\n(C) {a2}\n(D) {a3}\n(E) {a4}\nRespond with only the letter (A, B, C, D or E) of the correct option."
        video_path = os.path.join(
            args.data_path,
            "subset500_videos" if args.eval_type == "subset" else "mini_videos",
            f"{idx}.mp4",
        )

        video, speech, speech_lengths = load_video(
            video_path=video_path, audio_path=None, max_frames_num=32
        )
        speech = torch.stack([speech]).to(device).half()
        processor = model.get_vision_tower().image_processor
        processed_video = processor.preprocess(video, return_tensors="pt")["pixel_values"]
        image = [(processed_video, video[0].size, "video")]

        conv_template = "qwen_1_5"
        conv = copy.deepcopy(conv_templates[conv_template])
        conv.append_message(conv.roles[0], qs)
        conv.append_message(conv.roles[1], None)
        prompt_question = conv.get_prompt()
        
        parts = split_text(prompt_question, ["<image>", "<speech>"])
        input_ids = []
        for part in parts:
            if part == "<image>":
                input_ids.append(IMAGE_TOKEN_INDEX)
            elif part == "<speech>":
                input_ids.append(SPEECH_TOKEN_INDEX)
            else:
                input_ids.extend(tokenizer(part).input_ids)

        input_ids = torch.tensor(input_ids, dtype=torch.long).unsqueeze(0).to(device)
        image_tensor = [image[0][0].half()]
        image_sizes = [image[0][1]]
        generate_kwargs = {"eos_token_id": tokenizer.eos_token_id}

        with torch.inference_mode():
            cont = model.generate(
                input_ids,
                images=image_tensor,
                image_sizes=image_sizes,
                speech=speech,
                speech_lengths=speech_lengths,
                do_sample=False,
                temperature=0.5,
                max_new_tokens=5,
                modalities=["video"],
                **generate_kwargs,
            )
        pred = tokenizer.batch_decode(cont, skip_special_tokens=True)[0].strip()
        print(f"\n[Num-{idx} | Pred: {pred}")
        pred = pred.replace("Answer", "")

        letters = ["A", "B", "C", "D", "E"]
        pred_answer = re.findall("[\(\ ]*[A-E][\)\ ]*", pred)

        pred_answer = pred_answer[0].strip()
        pred_answer = pred_answer.strip("()")
        if pred_answer in letters:
            pred_idx = letters.index(pred_answer)
            pred = letters[pred_idx]
        else:
            print("pred_answer: ", pred_answer, " pred: ", pred, flush=True)
            pred_idx = 2
            pred = letters[pred_idx]

        outputs.append(
            {
                "idx": idx,
                "prompt": qs,
                "model_id": args.model_path,
                "video_name": video_name,
                "pred": pred_idx
            }
        )

    answers = json.load(
        open(
            os.path.join(
                args.data_path, "subset_answers.json" if args.eval_type == "subset" else "miniset_answers.json"
            ),
            "r",
        )
    )

    correct = 0
    total = 0
    for output in outputs:
        if output["idx"] in answers:
            output["answer"] = answers[output["idx"]]
            total += 1
            if int(output["pred"]) == int(answers[output["idx"]]):
                correct += 1

    print(f"Accuracy: {correct / total}")
    result = {"acc": correct / total}


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--model_path", type=str, default="/home/xxx/models/EgoGPT-7b-EgoIT-EgoLife"
    )
    parser.add_argument('--data_path', default="/home/xxx/datasets/egoschema")
    parser.add_argument('--eval_type', default="mini")
    args = parser.parse_args()
    main(args)

```

**Comment by choiszt** [comment]
Hi @Xiaolong-RRL,

The issue you're encountering arises because the multimodal tokens (`<image>\n` for images and videos, and `<speech>\n` for videos containing audio) are not correctly inserted into the context within the Egoschema prompt.

Currently, lmms-eval supports evaluation for both Egoschema and its subsets. To address your issue, please refer to:

[utils.py](https://github.com/EvolvingLMMs-Lab/lmms-eval/blob/main/lmms_eval/tasks/egoschema/utils.py) for detailed implementation related to Egoschema token handling.

[egogpt.py](https://github.com/EvolvingLMMs-Lab/lmms-eval/blob/e5f70443e19711e1ad225d60f412288f89c977cf/lmms_eval/models/egogpt.py#L427) (line 427 onwards) for practical examples of multimodal token insertion used by EgoGPT.

Let me know if you need further assistance!

**Comment by Xiaolong-RRL** [comment]
Awesome!! thanks so much for your quick and effective response!
